### ISSUE TITLE
[BUG] anvi-refine takes very long after update

### ISSUE THREAD

**Issue by AviHuene** [merenlab/anvio#1850]
## Short description of the problem

After update to v7.1 anvi-refine takes > 30 mins instead of ~2-5 mins to load interface.

## anvi'o version

```
Anvi'o .......................................: hope (v7.1)

Profile database .............................: 38
Contigs database .............................: 20
Pan database .................................: 15
Genome data storage ..........................: 7
Auxiliary data storage .......................: 2
Structure database ...........................: 2
Metabolic modules database ...................: 2
tRNA-seq database ............................: 2
```

## System info

System: Ubuntu 14.04
anvi'o install via conda.

## Detailed description of the issue

Further information on the dataset:
Size of Contigs. and Profile.db: 

```
num_contigs ..................................: 5761716
total_length .................................: 14475009184
num_splits ...................................: 5791598
```

I expected the refine interface to load in ~2-5 minutes, it now takes 30+ minutes or gets stuck at:
![anvi_refine_state](https://user-images.githubusercontent.com/66365465/145002330-3921299c-109d-4ad4-b768-1885aeaeb49b.png)

The example bin with 488 splits took 5 mins for the initialization of the databases. The following loading of views takes the longest, particularly for std_coverage. In most cases, it continues at some point to load up the interface. In this case its now been stuck at std_coverage for +1 hour (still running).

Thanks for the assistance!

**Comment by AviHuene** [comment]
Thank you for working on it immediately @meren. We have set up the dev environment now, so I should hopefully be able to help.

**Comment by meren** [comment]
Hey @AviHuene, I am now testing `anvi-refine` on a relatively large dataset (certainly not as large as yours),

```
num_contigs ..................................: 118,736
total_length .................................: 669,760,170
num_splits ...................................: 123,576
```

And I don't seem to be able to see any performance issue :(

Can you please run the following commands on your system and send me the output (you will need to replace `CONTIGS.db` and `PROFILE.db` with the paths of your own files):

``` bash
# select 488 random split names from the contigs database
time splits=`sqlite3 CONTIGS.db 'SELECT split FROM splits_basic_info ORDER BY RANDOM() LIMIT 400;' | awk '{print "\"" $0 "\""}'| paste -s -d ','`

# fetch data about them from the profile database:
time sqlite3 PROFILE.db "SELECT * FROM std_coverage_contigs where item IN ($splits);" > /dev/null
```

Meanwhile, I wonder @xvazquezc who had trouble with `anvi-refine` before and helped us improve it happen to use it in anvi'o `v7.1` and noticed any difference in performance .

**Comment by AviHuene** [comment]
Hi @meren,

I have run the commands (but am not sure it worked?).
Here is the output:
`sqlite3 CC_RP-merged/PROFILE.db  > /dev/null  0,00s user 0,00s system 93% cpu 0,005 total`

**Comment by meren** [comment]
Thanks for your quick response, @AviHuene, but it doesn't seem to have worked :)

Are you sure you copied the *entire* lines for both commands I sent you? Please copy-paste the entire output of your terminal (or send a screenshot).

**Comment by AviHuene** [comment]
I reran it a few times to be sure, but that's all I get @meren.

![anvio_select_splits](https://user-images.githubusercontent.com/66365465/145042247-969cdecd-c1d2-4007-9cac-b05bbc22ad24.png)

**Comment by AviHuene** [comment]
Hi @meren, I just found out, that our dev version is not working as it should. We will try and fix this, and then see if your commands work afterwards.
Sorry for the inconvenience :-/

**Comment by meren** [comment]
Hey @AviHuene, no worries at all. By the way, what's happening in your screenshot still doesn't make much sense to me :( This is how the output should look like:

![image](https://user-images.githubusercontent.com/197307/145052777-ac21d754-d493-4bab-9c9f-22e6d2889702.png)

I wonder if you are using `zsh` instead of `bash`. What does the output of this command tells you:

```
echo $SHELL
```

**Comment by AviHuene** [comment]
Hi @meren,
(we haven't fixed the installation yet), but it was indeed zsh and not bash.
The bash output is as follows:
![bash_output_splits](https://user-images.githubusercontent.com/66365465/145067866-63f72a86-6e7a-413e-b9b0-3c7b1623f443.png)

**Comment by meren** [comment]
Hey @AviHuene,

As you can see it takes only about 4 mins to read the data for std_coverages from the database, and is view is only one of five views. So only reading the view data will take about 20 mins :/ But important to point out, this step has nothing to do with anvi'o. Because in this test we simply are reading the data from that database file, so this exercise tells us more about the disk I/O of the system you are working with. So this tells me two things:

* It is impossible for the anvi'o interactive interface to load in less than 20 mins for these data on this system.
* It should not take more than 5 minutes to read a single view unless there are also memory issues on your system.

I will think more about this and see if there is a way to speed up the process by generating database indexes for these tables :/

**Comment by meren** [comment]
Hey @AviHuene,

Here is one more question for you that may speed things up quite a bit. If you can help me test this, perhaps I can implement it in anvi'o.

Please run these commands in this order (in a BASH terminal, and share the output, i.e., the entire terminal screen):

``` bash
# we create an index first
time sqlite3 CC_RP-merged/PROFILE.db 'CREATE INDEX std_coverage_contigs_idx ON std_coverage_contigs(item);'

# get random splits
time splits=`sqlite3 CC_RP.db 'SELECT split FROM splits_basic_info ORDER BY RANDOM() LIMIT 400;' | awk '{print "\"" $0 "\""}'| paste -s -d ','`

# recover data from them
time sqlite3 CC_RP-merged/PROFILE.db "SELECT * FROM std_coverage_contigs where item IN ($splits);" > /dev/null
```

**Comment by xvazquezc** [comment]
Hey @meren, I haven't run `anvi-refine` in quite a while (or anvi'o for that matter... ).
In any case, I'd ask @AviHuene if you are running `anvi-refine` in the exact same system as you did before 7.1, starting with the hard drive. With such an intensive I/O operation it makes a huge difference.

Also, @meren, something that I found after the prior issue with `anvi-refine` it was the I/O scheduler that my OS (Ubuntu 18.04 I think) had chosen at the time. It was affecting the HDD performance so badly that I actually ended replacing it, but it was even worse with intense I/O tasks.

**Comment by AviHuene** [comment]
Hey @meren, I have run the commands and hope this helps:

![anvio_sqlite_index](https://user-images.githubusercontent.com/66365465/145204806-0ce0b7c5-dbc8-4ddd-b8cd-7e3ff6d35a87.png)

@xvazquezc, thank you for the suggestion. As far as I know nothing changed between the updates, but we're investigating what it could be/ if something else is slowing things down.

**Comment by meren** [comment]
I can see that the creation of indexes for view tables using the item column will improved the performance of data retrieval even on your system about 500-fold, @AviHuene.

We really have to implement this ASAP. I'll work on it as soon as I can, and I hope you can test this new feature on the same system.

**Comment by AviHuene** [comment]
Sure thing, @meren. Let me know when/what I should test. 
Unfortunately, we are currently still struggling to get the dev version installed properly, as scikit is causing errors. The fixes posted in Issue #1682 have unfortunately not worked so far.

With version 0.24.2 profiling during the self test results in the following error:

:: Profiling sample SAMPLE-01 ...

Traceback (most recent call last):
  File "/home/dominik/ANALYSIS/opt/tools/anvio/anvio-dev/anvio/bin/anvi-profile", line 5, in <module>
    import anvio.profiler as profiler
  File "/home/dominik/ANALYSIS/opt/tools/anvio/anvio-dev/anvio/anvio/profiler.py", line 23, in <module>
    import anvio.clustering as clustering
  File "/home/dominik/ANALYSIS/opt/tools/anvio/anvio-dev/anvio/anvio/clustering.py", line 8, in <module>
    from sklearn import manifold
  File "/home/dominik/ANALYSIS/opt/tools/anvio/envs/anvio-dev/lib/python3.6/site-packages/sklearn/__init__.py", line 83, in <module>
    from .utils._show_versions import show_versions
  File "/home/dominik/ANALYSIS/opt/tools/anvio/envs/anvio-dev/lib/python3.6/site-packages/sklearn/utils/_show_versions.py", line 12, in <module>
    from .openmp_helpers import openmp_parallelism_enabled
ImportError: dlopen: cannot load any more object with static TLS 

Older versions of the scikit (e.g. v 0.20.) lead to the previously described "ModuleNotFoundError: No module named 'sklearn.ensemble._forest" and aborts at the interface step.

FYI: In the meantime I have tried a smaller dataset (451078 contigs, 1156188330 length, calling 175 splits) and this only took about 50s to load the interactive interface on v7.1. So it really is the big dataset causing trouble.

**Comment by xvazquezc** [comment]
@AviHuene I have `scikit-learn==0.23.1` in my `anvio-master` (dev) environment and it was working fine in case you want to try that version specifically
